File: src/main/GO/leetcode/lists/hot100/LeetCode221.go
package hot100

func maximalSquare(matrix [][]byte) int {
	if len(matrix) == 0 {
		return 0
	}
	dp := make([][]int, len(matrix))
	for i := 0; i < len(matrix); i++ {
		dp[i] = make([]int, len(matrix[0]))
	}
	var side = 0
	for i := 0; i < len(matrix[0]); i++ {
		if matrix[0][i] == '1' {
			dp[0][i] = 1
			side = 1
		}
	}
	for i := 0; i < len(matrix); i++ {
		if matrix[i][0] == '1' {
			dp[i][0] = 1
			side = 1
		}
	}
	for i := 1; i < len(matrix); i++ {
		for j := 1; j < len(matrix[0]); j++ {
			if matrix[i][j] == '1' {
				dp[i][j] = Min(Min(dp[i-1][j-1], dp[i-1][j]), dp[i][j-1]) + 1
			}
			side = Max(side, dp[i][j])
		}
	}
	return side * side
}

File: java/src/leetcode/problems/LeetCode221.java
package leetcode.problems;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/11/3 14:33
 */
public class LeetCode221 {
    public int maximalSquare(char[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        // dp[i][j] 表示以 (i, j) 为右下角的最大正方形边长
        int[][] dp = new int[m][n];
        int maxSide = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                // 只有当前位置是 '1' 才有可能构成正方形
                if (matrix[i][j] == '1') {
                    if (i == 0 || j == 0) {
                        // 边界情况：第一行或第一列，最大只能是 1
                        dp[i][j] = 1;
                    } else {
                        // 核心递推：受限于左、上、左上的最小值（短板效应）
                        dp[i][j] = Math.min(Math.min(dp[i - 1][j], dp[i][j - 1]), dp[i - 1][j - 1]) + 1;
                    }
                    maxSide = Math.max(maxSide, dp[i][j]);
                }
            }
        }
        return maxSide * maxSide;
    }
}
